libtenzir/builtins/functions/community_id.cpp

```diff
@@ -0,0 +1,251 @@
+};
+
+class plugin final : public function_plugin {
+public:
+  auto name() const -> std::string override {
+    return "tql2.community_id";
+  }
+
+  auto make_function(invocation inv, session ctx) const
+    -> failure_or<function_ptr> override {
+    auto args = arguments{};
+    TRY(argument_parser2::function("community_id")
+          .add("src_ip", args.src_ip)
+          .add("dst_ip", args.dst_ip)
+          .add("src_port", args.src_port)
+          .add("dst_port", args.dst_port)
+          .add("proto", args.proto)
+          .add("seed", args.seed)
+          .parse(inv, ctx));
+    auto startup_failure = false;
+    if (not args.src_ip) {
+      startup_failure = true;
+      diagnostic::error("missing required argument `src_ip`")
+        .primary(inv.call.fn)
+        .emit(ctx);
+    }
+    if (not args.dst_ip) {
+      startup_failure = true;
+      diagnostic::error("missing required argument `dst_ip`")
+        .primary(inv.call.fn)
+        .emit(ctx);
+    }
+    if (not args.proto) {
+      startup_failure = true;
+      diagnostic::error("missing required argument `proto`")
+        .primary(inv.call.fn)
+        .emit(ctx);
+    }
+    if (startup_failure) {
+      return failure::promise();
+    }
+    return function_use::make([args = std::move(args)](evaluator eval,
+                                                       session ctx) -> series {
+      auto null_series = [&] {
+        return series::null(string_type{}, eval.length());
+      };
+      auto src_ip_series = eval(*args.src_ip);
+      if (caf::holds_alternative<null_type>(src_ip_series.type)) {
+        return null_series();
+      }
+      auto dst_ip_series = eval(*args.dst_ip);
+      if (caf::holds_alternative<null_type>(dst_ip_series.type)) {
+        return null_series();
+      }
+      auto proto_series = eval(*args.proto);
+      if (caf::holds_alternative<null_type>(proto_series.type)) {
+        return null_series();
+      }
+      auto src_port_series = std::optional<series>{};
+      if (args.src_port) {
+        src_port_series = eval(*args.src_port);
+        if (caf::holds_alternative<null_type>(src_port_series->type)) {
+          src_port_series = std::nullopt;
+        }
+      }
+      auto dst_port_series = std::optional<series>{};
+      if (args.dst_port) {
+        dst_port_series = eval(*args.dst_port);
+        if (caf::holds_alternative<null_type>(dst_port_series->type)) {
+          dst_port_series = std::nullopt;
+        }
+      }
+      auto seed_series = std::optional<series>{};
+      if (args.seed) {
+        seed_series = eval(*args.seed);
+        if (caf::holds_alternative<null_type>(seed_series->type)) {
+          seed_series = std::nullopt;
+        }
+      }
+      auto src_ips = src_ip_series.as<ip_type>();
+      if (not src_ips) {
+        diagnostic::warning("`community_id` expected `ip` as 1st argument")
+          .primary(*args.src_ip)
+          .emit(ctx);
+        return null_series();
+      }
```

Comment:
Since they are now named, the "1st argument" doesn't make sense (same below). 

You can do this instead:

```cpp
diagnostic::error("expected argument of type `ip`, but got `{}`", src_ip_series.type.kind())
```

(Side note: We should probably factor this out at some point to provide templates for the common error messages)